### PR TITLE
Added wiggle png/gif modifier

### PR DESCRIPTION
**Pull request by MaxOhn**
Making pngs and gifs wiggle
  

Known bugs:

- The resulting gif of `[some png] -> [some gif modification] -> wiggle` is about twice as slow as `[some png] -> wiggle -> [some gif modification]`
- Very few gifs result in mysterious glitches e.g. 
For the majority of gifs it works properly though

### CODE REVIEW

File: gifmodify.js [hunk 1]
```diff
@@ -157,14 +158,76 @@ function addColorShiftingFramesGIF(inputGif, options, callback) {
...
+      [shift, left] = shiftStep(shift, left);
+    }
+    setFrameProperties(frames[i], { interlaced: false, bitmap: newFrame.bitmap });
+    // Set initial wiggle offset for next frame
+    [startShift, startLeft] = shiftStep(startShift, startLeft);
+  }
+  callback(frames);
+}
+
+function setFrameProperties(frame, options) {
```
Comment (Yentis) [inline gifmodify.js:198]: Add frame.interlaced = false to the top here as we never want this on and remove it from the function calls.

File: gifmodify.js [hunk 2]
```diff
@@ -157,14 +158,76 @@ function addColorShiftingFramesGIF(inputGif, options, callback) {
...
+        let codec = new GifCodec();
+        codec.encodeGif(frames).then(resultGif => {
+          resolve(resultGif.buffer);
+        });
+      });
+    }).catch(error => reject(error));
+  });
+}
+
+function addWigglingFramesGIF(inputGif, options, callback) {
```
Comment (Yentis) [inline gifmodify.js:180]: This function is very similar to addWigglingFramesPNG, can you put this in a separate function like I did with rainbow?

File: gifmodify.js [hunk 3]
```diff
@@ -157,14 +158,76 @@ function addColorShiftingFramesGIF(inputGif, options, callback) {
...
+  let width = imgWidth + 2 * Math.floor(imgWidth / 15); // ~6.6% of width is wiggle room for both sides
+  let margin = width - imgWidth;
+
+  let shiftSize = Math.max(1, Math.floor(margin / 6));
+  let interval = 2 * (margin / shiftSize + 4);
+  let stripeHeight = 2 * shiftSize;
+  let frames = alignGif(inputGif.frames, interval);
+
+  let startShift = margin / 2; // Initial offset of wiggle
+  let startLeft = true; // true -> go to left
```
Comment (Yentis) [inline gifmodify.js]: This is always true, make it an option?

File: gifmodify.js [hunk 4]
```diff
@@ -157,14 +158,76 @@ function addColorShiftingFramesGIF(inputGif, options, callback) {
...
+
+  function shiftStep(s, l) {
```
Comment (Yentis) [inline gifmodify.js]: why is this inside the addWigglingFramesGIF function?

File: gifmodify.js [hunk 5]
```diff
@@ -333,6 +393,97 @@ function shiftColor(imgData, index, shiftAmount, randomBlack, randomWhite) {
   return colors;
 }
 
+exports.createWigglingPNG = function(options) {
+  return new Promise((resolve, reject) => {
+    loadImage(options.buffer).then(image => {
+      let size = options.isResized ? 1 : options.size;
+      let height = Math.floor(size * image.height);
```
Comment (Yentis) [inline gifmodify.js]: change this in preparePNGVariables and then you can use that function here too

File: gifmodify.js [hunk 6]
```diff
@@ -333,6 +393,97 @@ function shiftColor(imgData, index, shiftAmount, randomBlack, randomWhite) {
...
+      width += 2 * Math.floor(width / 15); // ~6.6% of width is wiggle room for both sides
+
+      let encoder = new GIFEncoder(width, height);
+      getBuffer(encoder.createReadStream()).then(buffer => resolve(buffer));
+      setEncoderProperties(encoder, options.value * 3);
+
+      addWigglingFramesPNG({
+        image: image,
+        width: width,
+        imgWidth: Math.floor(size * image.width),
```
Comment (Yentis) [inline gifmodify.js]: You're doing this twice, rename line 401 to imgWidth and make 402 width. Then change imgWidth here to imgWidth: imgWidth

File: gifmodify.js (hunk 3 above)
Comment (MaxOhn) [reply gifmodify.js]: Starting off left or right only offsets the wiggle, still looks very similar so it wouldn't really matter
Start left:  Start right:

**Pull request merged**